File: tdd/password-validator/src/PasswordValidator.php
<?php declare(strict_types=1);

namespace Kata;

use Kata\Rules\PasswordRuleInterface;

final class PasswordValidator
{
    /**
     * @var list<PasswordRuleInterface>
     */
    private array $passwordRules;

    public function __construct(array $passwordRules)
    {
        $this->passwordRules = $passwordRules;
    }

    public function validate(string $password): bool
    {
        foreach ($this->passwordRules as $rule) {
            if (!$rule->validateRule($password)) {
                return false;
            }
        }

        return true;
    }
}

File: tdd/password-validator/kotlin/src/main/kotlin/PasswordValidator.kt
class PasswordValidator {

    fun validate(password: String): Boolean {
        return password.length >= 9
            && password.any { it.isUpperCase() }
            && password.any { it.isLowerCase() }
            && password.any { it.isDigit() }
            && password.any { it == '_' }
    }
}
